internal/worker/events.go

```diff
@@ -0,0 +1,172 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/webhooks"
+	"go.uber.org/zap"
+)
+
+type (
+	EventManager interface {
+		AddSubscriber(id string, s EventSubscriber) (chan struct{}, error)
+		HandleEvent(event webhooks.Event)
+		Run(ctx context.Context, eventURL string, opts ...webhooks.HeaderOption) error
+		Shutdown(context.Context) error
+	}
+
+	EventSubscriber interface {
+		HandleEvent(event webhooks.Event) error
+		Subscribe(e EventManager) error
+	}
+
+	WebhookManager interface {
+		RegisterWebhook(ctx context.Context, wh webhooks.Webhook) error
+		UnregisterWebhook(ctx context.Context, wh webhooks.Webhook) error
+	}
+)
+
+type (
+	eventManager struct {
+		webhooks WebhookManager
+		logger   *zap.SugaredLogger
+
+		registerInterval time.Duration
+
+		mu         sync.Mutex
+		subs       map[string]eventSubscriber
+		registered []webhooks.Webhook
+	}
+
+	eventSubscriber struct {
+		EventSubscriber
+		readyChan chan struct{}
+	}
+)
+
+func NewEventManager(w WebhookManager, l *zap.Logger, registerInterval time.Duration) EventManager {
+	return &eventManager{
+		webhooks: w,
+		logger:   l.Sugar().Named("events"),
+
+		registerInterval: registerInterval,
+
+		subs: make(map[string]eventSubscriber),
+	}
+}
+
+func (e *eventManager) AddSubscriber(id string, s EventSubscriber) (chan struct{}, error) {
```

Comment:
I get the motivation behind a generic `eventManager` but the way it is right now feels a bit unfinished. Tbh I wouldn't have minded running the `cache.Initialize` in a loop instead of adding the `EventManager` abstraction but since you added it I got a few thoughts.

`AddSubscriber` adds subscribers but the subscribers can't define what webhooks they are interested in. So the logic of required webhooks is separated from the subscriber that consumes the events. However, if we changed it, `AddSubscriber` would need to replace `Run` altogether. Since every call to `AddSubscriber` would check that all webhooks the subscriber is interested in is registered and otherwise register it. So essentially `AddSubscriber` would spawn a loop and return a channel that the subscriber can check.

Reply:
I'm excited about the event manager because I don't like how (I) implemented the worker cache issue. Pulling it out of the worker cache should in time make the cache obsolete because things like the upload and download manager could load their state when they're initialised and then just handle events from the bus.

We can change it of course, but I don't immediately see the upside in adding fine-grained control over what events subscribers subscribe to. Instead, subscribers just subscribe and "see" all events, in their `HandleEvent` they have the option to ignore whatever they want to ignore. This version was my attempt to keep it simple, it's easy to change afterwards and I feel it would only add complexity at this point. Fine with changing it though.

Reply:
That's fine. I just didn't fully understand the way you were going with this. In some ways it seemed like you wanted fine-grained control (e.g. the channel per subscriber) and in other ways it didn't (e.g. having the event manager subscribe to everything and pass it on).